### ISSUE TITLE
Improve docs on how to write function adaptors and how to use forced_parameters

### ISSUE THREAD

**Issue by ldivenere** [iminuit/iminuit#180]
Hi,
I'm trying to use it to minimize a function for which I don't know how many free paramenters I have a priori. I define my python fcn using keyword arguments. Then, I separate the different cases during the fcn evaluation.

First, in this case iminuit is not able to detect the parameter names, which is easily solved using the "forced_parameters" argument.
However, it seems that iminuit does not call my fcn function properly, since I get this error:

I guess this is because it tries to call it passing positional arguments (one for each paramenter), while my fcn doesn't have any positional argument in its definition.

Here it is a simplified code of how my fcn works:

```
import iminuit
import numpy as np
x=np.arange(10)
y=2*x+1
err_y=y*0.1

def fcn(**params): ##just an example
    c1=params["c1"]
    c2=params["c2"]
    chi2= ((y-c1*x-c2)/err_y)**2
    return chi2.sum()

## define the dictionary for iminuit parameters
kw={}
kw["c1"]=1.9
kw["c2"]=0.9
kw["error_c1"]=0.1
kw["error_c2"]=0.1
min=iminuit.Minuit(fcn, errordef=1, forced_parameters=["c1","c2"], **kw)
min.migrad()

TypeError: fcn() takes exactly 0 arguments (2 given)
```

This is how fcn must be correctly called:
`fcn(c1=1.9,c2=0.9)`

If called with positional arguments it raises the same TypeError:

```
fcn(1.9,0.9)

TypeError: fcn() takes exactly 0 arguments (2 given)
```

Looking through the code I found the possible reason in file PythonFCN.h line 97:
`PyObject* result = PyObject_Call(fcn,tuple,NULL);`
Maybe, the solution could be to pass the arguments as keyword arguments through a dictionary, with something similar to this
`PyObject* result = PyObject_Call(fcn,NULL,dict);`
where dict is a PyDict object which must be properly constructed.
I'm not a C++ expert, I couldn't make the code work properly.

Can anyone help? Do you think this feature could be implemented?
Leonardo

**Comment by jimcost** [comment]
Remove the ** from the function definition I recommend.

Also, if you dig around - you'll realize (1) the probfit module was created by the iminuit people to create cost functions and (2) if you want to define your own function the code will look something like this - 

# Chi-squared object for iminuit minimization
class Chi2:
    def __init__(self, f, x, y, var):
        self.f, self.x, self.y, self.var = (f,x,y,var)
        fSig = iminuit.describe(f) 
        self.func_code = make_func_code(fSig[1:])
        self.func_defaults = None # keeps np.vectorize happy
    def __call__(self, *arg):
        return sum((y-self.f(x,*arg))**2/var for x,y,var in zip(self.x, self.y, self.var))

I did not create this code, but found it on the internet and adapted it.

**Comment by ldivenere** [comment]
Hi Jim,

thanks for your help. My example was a very simplified one to point out the issue.
Your solution does work and it is what I did to avoid the issue, even though it does not solve the problem. I describe it in detail just in case someone else faces the same problem.
The fcn is defined with positional arguments instead of keyword arguments (`*params` instead of `**params` ). In this case, the object `params` is a python tuple (instead of a dictionary) and I can recall the parameters only through an integer index, which is not very handy when you have a lot of different parameters.

This is the updated version of the code I wrote before.
```
def fcn(*params): ##just an example
    c1=params[0]
    c2=params[1]
    chi2= ((y-c1*x-c2)/err_y)**2
    return chi2.sum()
```
The order in which iminuit passes the parameters to the fcn is the same as in the `forced_parameters` list. So, in this case the iminuit definition must be:
```
iminuit.Minuit(fcn, errordef=1, forced_parameters=["c1","c2"], **kw)
```

Leonardo

**Comment by cdeil** [comment]
@ldivenere - Indeed, `iminuit` does not support a keyword argument based callback function interface.

If you have such a function, you should write a wrapper function or class with `__call__` and pass that to `Minuit`, as in the example above. We did something similar [here](https://github.com/gammapy/gammapy/blob/d1b94474ac93e256b4b8d9a85f373a531b7cda84/gammapy/utils/fitting/iminuit.py#L50).

Recently @HDembinski added a second callback function interface, described as option b) here:
http://iminuit.readthedocs.io/en/latest/api.html#iminuit.Minuit
It is not keyword argument based, like what you had though, you'd still need a wrapper function to use your keyword-argument based method with `iminuit`.

@HDembinski - I'm assigning this issue to you and the v1.3 milestone.
I'm very much -1 on adding more function interfaces to `iminuit`, but clearly the docs need to be more clear on this.
My suggestion would be to consider to move the content from http://iminuit.readthedocs.io/en/latest/api.html#iminuit.Minuit describing hte callback functions and how to write a wrapper if needed on a separate docs page, and to link to that docs page from the class docstring. I'd also be happy to write such a docs page, so feel free to assign over to me if you prefer.

**Comment by HDembinski** [comment]
👍 on writing adding an example to the docs which shows how to write simple wrappers.
I am also 👎 on adding keyword argument based callback function to Minuit. This is a rather obscure way of defining a function. Luckily, writing simple adapters with lambdas is very easy.

```
def fcn(**params): ## example from OP
    c1=params["c1"]
    c2=params["c2"]
    chi2= ((y-c1*x-c2)/err_y)**2
    return chi2.sum()

# use a lambda adaptor
m = Minuit(lambda c1, c2: fcn(c1=c1, c2=c2), c1=1.9, c2=0.9, error_c1=0.1, error_c2=0.1)
m.migrad()
```

**Comment by HDembinski** [comment]
@cdeil Please do not move any content from the doc strings in http://iminuit.readthedocs.io/en/latest/api.html#iminuit.Minuit, the in-place docs there are good to have. I think we should merely write some high-level docs to explain the allowed function signatures with some examples.

**Comment by HDembinski** [comment]
This is addressed in PR #251
